draft-pauly-taps-arch.md

```diff
@@ -205,103 +205,93 @@ Beyond the basic objects, there are several high-level groups of actions that an
 
->> A Content Sent event notifies the application of the status of its Send action. This may be an error, an indication that Content has been processed by the protocol stack, or potentially \[MICHAEL: Why "potentially" - because you're not sure about it? If it may be such an indication, "potentially" should be removed (this seems like the right thing to do for me).] an indication that Content has been acknowledged by a peer.
+* Connection Finished: A Connection Finished event signals to an application that a given Connection is no longer usable for sending or receiving Content. This should deliver a useful error to the application.
 
-Path Properties Changed
+* Connection Received: A Connection Received event signals to an application that a given Listener has passively received a Connection.
 
->> A Path Properties Changed event notifies the application that some property of the Connection has changed that may influence how and where data is sent and/or received.
+* Content Received: A Content Received event delivers received content to the application, based on a Receive action. This may include an error if the action failed.
 
-### Termination {#termination}
+* Content Sent: A Content Sent event notifies the application of the status of its Send action. This may be an error, an indication that Content has been processed by the protocol stack, or potentially \[MICHAEL: Why "potentially" - because you're not sure about it? If it may be such an indication, "potentially" should be removed (this seems like the right thing to do for me).] an indication that Content has been acknowledged by a peer.
 
-Close
+* Path Properties Changed: A Path Properties Changed event notifies the application that some property of the Connection has changed that may influence how and where data is sent and/or received.
 
->> Close is the action an application may take on a Connection to indicate that it no longer intends to send data, is no longer willing to receive data, and that the protocol should signal this state to the remote endpoint if applicable.
+### Termination Actions {#termination}
 
-Abort
+* Close() is the action an application may take on a Connection to indicate that it no longer intends to send data, is no longer willing to receive data, and that the protocol should signal this state to the remote endpoint if applicable.
 
->> Abort is an action the application may take on a Connection to indicate a Close, but with the additional indication that the transport system should not attempt to deliver any outstanding data.
+* Abort() is an action the application may take on a Connection to indicate a Close, but with the additional indication that the transport system should not attempt to deliver any outstanding data.
 
```

Comment:
Should there also be an action to indicate that the application will not send any further data but is still willing to receive data?

Reply:
No; this is explained in minset - exactly why I hesitate to move stuff to the appendix, people don't seem to read the appendix   :)    if you want to support a broad set of transport connections, you need these primitives to only expose the most limiting semantics of all of them. What you describe is a TCP specialty, it's the result of half-closed connections. If you expose this, and someone wants to use it, you limit the system to only support protocols that can do half-closed connections. Worse, you need to initially ask "will you want to use the primitive that still lets you receive after closing?" in order to choose the right set of protocols (or simply never use a protocol that can't do half-closed connections, which are by themselves not very useful anyway).

Reply:
I agree with @mwelzl that this action is of limited utility for non-TCP stacks, and we could easily leave this out for London, but in general I don't think it's a problem to add actions which express application intent but which some/many stacks will treat as a no-op. I see `EndOfData()` (it's hard to name this, cf. socket's terrible `shutdown()`) as precisely this kind of thing -- it expresses an application-layer semantic, but some/many/most protocols might ignore it, and that's also okay as long as it doesn't cause incorrect operation of those protocols. (UDPLite's "checksum coverage" is a perfect example of this -- it's useless in TCP, and useless in UDP unless it's 0, but it doesn't express "don't checksum this", it expresses "the application is tolerant of corruption of this data". Same same.)

Reply:
No sorry, this is different: if you write an application that sends data where it may be fine to not use a checksum, but you get a checksum, nothing bad happens. If you write an application where you issue close but you do expect to be able to read afterwards, but you can't, this is semantically different. Application 1 must ensure that communication is really over and it doesn't expect anything anymore from app 2 when the it closes - otherwise, the sent data from app2 is lost. So, at the end of the connection, you're in fact unreliable.

This isn't a big problem in practice, and that's why TCP's half-closed connections are somewhat obscure - but if you rely on them being there, your application can break. If you don't get "checksum disabled", nothing will ever break.

Reply:
yep, agreed, that's why it's a different action: `EndOfData()`, not `Close()`. Also agree that `Close()` should mean the connection should be torn down and no further reads or writes are expected (with the caveat, of course, that in an asynchronous API the API cannot guarantee that no `Received<>` event will fire after `Close()` is called).

Reply:
(and I should point out that you still have to `Close()` after `EndOfData()`... it's all a bit obscure though, so I'd be happy to drop it for London)

Reply:
Oh sure, that can happen. No problem though. The specific rules we had written up about this in NEAT for an active close were: "Triggered by calling neat_shutdown();
on_writable() will never be called again, calling neat_write() returns error
on_readable() can be called"

Reply:
Thinking more about this - perhaps we should add this information (about the read-callback potentially still firing) to the description of the call?  I left it out on purpose to avoid making this text unnecessarily complex, but if you think it helps then we could include this, for "close" and "abort" ?